### DISCUSSION TITLE
自定义组件在透明背景的container容器中删除后图像未消失

### DISCUSSION THREAD

**Discussion by BowinChow** [VincentWei/minigui#104] [Q&A]
**自定义组件的思路：**
  根据MiniGUI绘制图片/文字/矩形的方法，可以将图片，文字以及透明框的矩形放到一个panelPiece中。通过调整不同piece的相对位置可以组合为类似安卓toast的组件效果。
![image](https://user-images.githubusercontent.com/52844652/200983892-d5130fbf-a196-43d1-962a-ed3dbb5801bd.png)
但是需要注意的是，toast的出现与消失是异步的（用户或场景决定什么时机出现）。那么如何在基于MiniGUI的cell-ux-demo中实现呢？我的思路就是利用了MiniGUI中的PostMessage方法，即当触发绘制组件时延时一段时间再删除对应组件。具体代码为：

`#define MSG_REFRESH_UI 100054
#define FONT_SIZE 50
#define TOAST_LENGTH_LONG 5
#define TOAST_LENGTH_SHORT 3
#define TOAST_BORDER_RADIUS 20
#define TOAST_RECT_BORDERSIZE 6

#define TOAST_RECT_BORDER_COLOR_R 0xff
#define TOAST_RECT_BORDER_COLOR_G 0x00
#define TOAST_RECT_BORDER_COLOR_B 0x00
#define TOAST_RECT_BORDER_COLOR_A 0x65

#define TOAST_RECT_COLOR_R 0xff
#define TOAST_RECT_COLOR_G 0x00
#define TOAST_RECT_COLOR_B 0x00
#define TOAST_RECT_COLOR_A 0x02

typedef struct _GradientData {
    float pos;
    ARGB color;
} GradientData;
static const GradientData infobox1_gradient[] = {
        {0.00f, 0xff000000,},
        {1.00f, 0xff000000,},
};
void MakeToast::postRefreshUIMessage(mWidget *self) {
    sleep(TOAST_LENGTH_SHORT);
    if (ToastManager::getInstance()->threadStatus == ToastManager::getInstance()->THREAD_STOP) {
        LOGD(LOG_AR_MEDIA, "stop postMessage thread success!");
        return;
    }
    PostMessage(self->hwnd, MSG_REFRESH_UI, (WPARAM) self, 0);
}

void MakeToast::onRefreshUI(mWidget *self, int message, WPARAM wparam, LPARAM lparam) {
    ToastManager::getInstance()->deletePieceCache(self);
}

MakeToast *MakeToast::MakeToast = NULL;

MakeToast *MakeToast::getInstance() {
    if (MakeToast == NULL) {
        MakeToast = new MakeToast();
    }
    return MakeToast;
}

void MakeToast::destroyInstance() {
    if (MakeToast) {
        delete MakeToast;
        MakeToast = nullptr;
    }
}`

`void MakeToast::showToast(int width, int height, char *imgPath, DWORD str, DWORD color, mWidget *actWidget) {

    LOGD(LOG_AR_MEDIA, "get toast");
    mPanelPiece *content = ToastManager::getInstance()->getPanelPiece();
    ToastManager::getInstance()->mTextFont = CreateLogFontEx("ttf", "helvetica", "UTF-8",
                                                            FONT_WEIGHT_MEDIUM,
                                                            FONT_SLANT_ROMAN,
                                                            FONT_FLIP_NONE,
                                                            FONT_OTHER_NONE,
                                                            FONT_DECORATE_NONE, FONT_RENDER_SUBPIXEL,
                                                            FONT_SIZE, 0);

    int toastWidth = width;
    int toastHeight = height;
    int toastPosX = (ACTIVITY_W - toastWidth) / 2;
    int toastPosY = 10;

    int imgWidth = 200;
    int imgHeight = 200;

    int offset = -40;

    int textWidth = 400;
    int textHeight = 200;

    mHotPiece *back_gradient = (mHotPiece *) NEWPIECE(mShapeTransRoundPiece);
    ToastManager::getInstance()->setBackGradientPiece(back_gradient);
    RECT rect_Bg;
    SetRect(&rect_Bg, 0, 0, 600, 200);
    _c(back_gradient)->setRect(back_gradient, &rect_Bg);
    _c(back_gradient)->setProperty(back_gradient, NCSP_TRANROUND_FILLMODE, NCSP_TRANROUND_GRADIENT_FILL);
    _c(back_gradient)->setProperty(back_gradient, NCSP_TRANROUND_CORNERFLAG,
                                   TRANROUND_CORNERFLAG_LEFTTOP | TRANROUND_CORNERFLAG_RIGHTTOP |
                                   TRANROUND_CORNERFLAG_RIGHTBOTTOM | TRANROUND_CORNERFLAG_LEFTBOTTOM);
    _c(back_gradient)->setProperty(back_gradient, NCSP_TRANROUND_RADIUS, 0);

    HBRUSH brush = _c((mShapeTransRoundPiece *) back_gradient)->getBrush(
            (mShapeTransRoundPiece *) back_gradient);
    for (unsigned int i = 0; i < TABLESIZE(infobox1_gradient); ++i)
        MGPlusLinearGradientBrushAddColor(brush, infobox1_gradient[i].color, infobox1_gradient[i].pos);
    _c(content)->addContent(content, back_gradient, (ACTIVITY_W -600)/2, 10);

    ToastManager::getInstance()->setRectPiece((mHotPiece *) NEWPIECE(mRectPiece));
    mHotPiece *rectPiece = ToastManager::getInstance()->getRectPiece();
    RECT rect_rec;
    SetRect(&rect_rec, 0, 0, toastWidth, toastHeight);
    _c(rectPiece)->setRect(rectPiece, &rect_rec);
    _c(rectPiece)->setProperty(rectPiece, NCSP_RECTPIECE_BORDERCOLOR, MakeRGBA(TOAST_RECT_BORDER_COLOR_R, TOAST_RECT_COLOR_G, TOAST_RECT_COLOR_B, TOAST_RECT_BORDER_COLOR_A));
    _c(rectPiece)->setProperty(rectPiece, NCSP_RECTPIECE_FILLCOLOR, MakeRGBA(TOAST_RECT_COLOR_R, TOAST_RECT_COLOR_G, TOAST_RECT_COLOR_B, TOAST_RECT_COLOR_A));
    _c(rectPiece)->setProperty(rectPiece, NCSP_RECTPIECE_YRADIUS, TOAST_BORDER_RADIUS);
    _c(rectPiece)->setProperty(rectPiece, NCSP_RECTPIECE_XRADIUS, TOAST_BORDER_RADIUS);
    _c(rectPiece)->setProperty(rectPiece, NCSP_RECTPIECE_BORDERSIZE, TOAST_RECT_BORDERSIZE);
    _c(content)->addContent(content, rectPiece, toastPosX, toastPosY);

    ToastManager::getInstance()->setTextPiece((mHotPiece *) NEWPIECE(mTextPiece));
    mHotPiece *textPiece = ToastManager::getInstance()->getTextPiece();
    RECT rect;
    SetRect(&rect, 0, 0, 400, 200);
    _c(textPiece)->setRect(textPiece, &rect);
    _c(textPiece)->setProperty(
            textPiece,
            NCSP_TEXTPIECE_LOGFONT, (DWORD)(ToastManager::getInstance()->mTextFont));
    _c(textPiece)->setProperty(
            textPiece,
            NCSP_LABELPIECE_LABEL, str);
    _c(textPiece)->setProperty(
            textPiece,
            NCSP_LABELPIECE_ALIGN, (DWORD) NCS_ALIGN_CENTER);
    _c(textPiece)->setProperty(
            textPiece,
            NCSP_TEXTPIECE_TEXTCOLOR, color);
    _c(content)->addContent(content, textPiece, (toastWidth-imgWidth)/2 + toastPosX, toastPosY + toastHeight / 2 - 100);

    ToastManager::getInstance()->setImagePiece((mHotPiece *) NEWPIECE(mImagePiece));
    mHotPiece *imgPiece = ToastManager::getInstance()->getImagePiece();
    BITMAP * bitmap = (BITMAP *) Load32Resource(imgPath, RES_TYPE_IMAGE, (DWORD) HDC_SCREEN);
    if (bitmap) {
        RECT rect;
        SetRect(&rect, 0, 0, imgWidth, imgHeight);
        _c(imgPiece)->setRect(imgPiece, &rect);
        _c(imgPiece)->setProperty(imgPiece, NCSP_IMAGEPIECE_IMAGE, (DWORD) bitmap);
    }
    _c(content)->addContent(content, imgPiece, toastPosX, toastPosY + toastHeight / 2 - 100);

    ncsSetComponentHandler((mComponent *) actWidget, MSG_REFRESH_UI, (void *) onRefreshUI);
    ToastManager::getInstance()->task = thread(MakeToast::postRefreshUIMessage, actWidget);
    ToastManager::getInstance()->task.detach();
    InvalidateRect(actWidget->hwnd,NULL,TRUE);

}
`

其中消失的逻辑为：

`void MakeToast::onRefreshUI(mWidget *self, int message, WPARAM wparam, LPARAM lparam) {
    ToastManager::getInstance()->deletePieceCache(self);
}`

`void ToastManager::deletePieceCache(mWidget *self) {
    _c(panelPiece)->delContent(panelPiece, imagePiece);
    _c(panelPiece)->delContent(panelPiece, textPiece);
    _c(panelPiece)->delContent(panelPiece, rectPiece);
    PanelPiece_invalidatePiece((mHotPiece *) panelPiece, nullptr);
    InvalidateRect(self->hwnd, NULL, TRUE);
    DestroyLogFont(mTextFont);
}`

当然我们需要事先把绘制的panelpiece在activity中注册到数据类：ToastManager中。
经过测试以上逻辑在容器背景不透明时是没问题的。但是如果将activity中的container的背景设置为WS_EX_TRANSPARENT则组件消失失败，经排查日志发现所有流程均和非透明时一致，尝试刷新window或者其他方法都不能解决问题，想请问广大的开发者们对这一问题有什么看法与解决办法吗？该场景比较好复现。

`static NCS_WND_TEMPLATE _ctrl_tmpl[] = {
    {NCSCTRL_CONTAINERCTRL,
     IDC_CONTAINER,
     ACTIVITY_X, ACTIVITY_Y, ACTIVITY_W, ACTIVITY_H,
     WS_VISIBLE,
     WS_EX_TRANSPARENT,
     "Container",
     NULL, NULL, NULL, NULL, 0, 0},
};
`

值得注意的是：
**如果按照上述代码添加一个完全不透明的背景（mShapeTransRoundPiece），那么自定义组件会消失，但是组件消失后会留下背景色，并且这时立即删除背景，界面上还会存在背景颜色。我怀疑是背景再刷新时遮盖了toast组件。**
